File: client/src/api/protocols.ts
import { apiFetch } from "./client";

export interface ProtocolVersion {
  id: string;
  protocol_id: string;
  version: number;
  created_at: string;
}

export interface Protocol {
  id: string;
  code: string;
  title: string;
  category: string | null;
  duration: string | null;
  body_html: string | null;
  version: number;
  tags: string[];
  is_favorite: boolean;
  author_name: string | null;
  created_at: string;
  updated_at: string;
  versions: ProtocolVersion[];
}

export interface ProtocolCreate {
  title: string;
  category?: string | null;
  duration?: string | null;
  body_html?: string | null;
  tags?: string[];
  author_name?: string | null;
}

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    const detail = body?.detail;
    if (Array.isArray(detail)) {
      throw new Error(detail.map((e: { loc?: string[]; msg?: string }) =>
        `${e.loc?.slice(-1)[0] ?? "champ"}: ${e.msg ?? "erreur"}`).join(", "));
    }
    throw new Error(typeof detail === "string" ? detail : `Erreur ${res.status}`);
  }
  if (res.status === 204) return undefined as T;
  return res.json();
}

export const listProtocols = (q = "", category?: string | null, favorites?: boolean) => {
  const p = new URLSearchParams({ q });
  if (category) p.set("category", category);
  if (favorites) p.set("favorites", "true");
  return apiFetch(`/api/protocols?${p}`).then((r) => json<Protocol[]>(r));
};

export const getProtocol = (id: string) =>
  apiFetch(`/api/protocols/${id}`).then((r) => json<Protocol>(r));

export const createProtocol = (body: ProtocolCreate) =>
  apiFetch("/api/protocols", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  }).then((r) => json<Protocol>(r));

export const updateProtocol = (id: string, body: Partial<ProtocolCreate> & { is_favorite?: boolean }) =>
  apiFetch(`/api/protocols/${id}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  }).then((r) => json<Protocol>(r));

export const deleteProtocol = (id: string) =>
  apiFetch(`/api/protocols/${id}`, { method: "DELETE" }).then((r) => json<void>(r));

export const restoreVersion = (id: string, version: number) =>
  apiFetch(`/api/protocols/${id}/restore/${version}`, { method: "POST" }).then((r) => json<Protocol>(r));

File: server/app/schemas/protocols.py
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class ProtocolVersionRead(BaseModel):
    id: UUID
    protocol_id: UUID
    version: int
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ProtocolCreate(BaseModel):
    title: str
    category: Optional[str] = None
    duration: Optional[str] = None
    body_html: Optional[str] = None
    tags: list[str] = []
    author_name: Optional[str] = None


class ProtocolUpdate(BaseModel):
    title: Optional[str] = None
    category: Optional[str] = None
    duration: Optional[str] = None
    body_html: Optional[str] = None
    tags: Optional[list[str]] = None
    author_name: Optional[str] = None
    is_favorite: Optional[bool] = None


class ProtocolRead(BaseModel):
    id: UUID
    code: str
    title: str
    category: Optional[str]
    duration: Optional[str]
    body_html: Optional[str]
    version: int
    tags: list[str]
    is_favorite: bool
    author_name: Optional[str]
    created_at: datetime
    updated_at: datetime
    versions: list[ProtocolVersionRead] = []
    model_config = ConfigDict(from_attributes=True)


class ProtocolListItem(BaseModel):
    id: UUID
    code: str
    title: str
    category: Optional[str]
    duration: Optional[str]
    version: int
    tags: list[str]
    is_favorite: bool
    author_name: Optional[str]
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)
